### ISSUE TITLE
When set SSL_SESS_CACHE_OFF in server side, if using TLS 1.3, server report fatal error: ssl_get_prev_session:session id context uninitialized

### ISSUE THREAD

**Issue by veronicafish** [openssl/openssl#10168]
Hi,

I have a server which depends on OPENSSL for SSL connection.

Everything works fine in TLS 1.2 and before. After upgrade to OPENSSL 1.1.1 and using TLS 1.3, server report: ssl_get_prev_session:session id context uninitialized.

I dig into it, find it is related to session resumption, and in server code, we set SSL_SESS_CACHE_OFF. I don't find any other special setting in server code.

As far as I understand, SSL_SESS_CACHE_OFF should only effect stateful ticket, not stateless ticket.

I want to know how it changes in OPENSSL 1.1.1, and why the error shows in TLS 1.3?

**Comment by mattcaswell** [comment]
That error is raised here:

https://github.com/openssl/openssl/blob/df22cbb555430b7206a8d30cb41f4e28b9e28370/ssl/ssl_sess.c#L566-L581

Have you changed the default verify_mode through a call to `SSL_CTX_set_verify()` or `SSL_set_verify`? And have you called `SSL_CTX_set_session_id_context()` or `SSL_set_session_id_context()`.

Prior to your upgrade to 1.1.1, which version of OpenSSL were you using? I can't think why you would be seeing different behaviour in 1.1.1 vs earlier versions. As far as I remember this check has been there for a long time.

**Comment by veronicafish** [comment]
Hi Mattcaswell,

Thank you for your respond. 

1. We do set verify_mode using:
SSL_CTX_set_verify(sslContext,
                SSL_VERIFY_PEER | SSL_VERIFY_CLIENT_ONCE, prepareForCallback);

2. We don't call SSL_CTX_set_session_id_context() or SSL_set_session_id_context().
As far as I understand, if set SSL_SESS_CACHE_OFF, then it means disable session ID resumption, so don't need to set session_id_context, am I right?

3. Our server works for many years, so we already used many openssl version, like 1.0.2, 1.0.1, 0.9.8, etc.
And I think it doesn't relate to OPENSSL 1.1.1 itself, it relates to TLS 1.3. Because I did test, after upgrade to OPENSSL 1.1.1, server works fine when using TLS 1.2, but fails on TLS 1.3.

**Comment by mattcaswell** [comment]
> As far as I understand, if set SSL_SESS_CACHE_OFF, then it means disable session ID resumption, so don't need to set session_id_context, am I right?

`SSL_SESS_CACHE_OFF`, when called on the server, disables caching of server side sessions. It does *not* switch off resumption. Resumption can still occur if a stateless session ticket is used (even in TLSv1.2). If resumption occurs and the verify mode has `SSL_VERIFY_PEER` set then a session_id_context must be set otherwise you will get this error.

The error itself has been present for a long time. I just did some tests with OpenSSL 1.0.2. I edited the source for s_server and commented out all calls to SSL_set_session_id_context() or SSL_CTX_set_session_id_context(). Then I invoked s_server like this:

````
openssl s_server -no_cache -Verify 5
````

This has the effect of setting `SSL_SESS_CACHE_OFF` and the verify mode to `SSL_VERIFY_PEER`.

Next I invoked s_client to establish a session:

````
openssl s_client -sess_out sess.pem -cert cert.pem -key key.pem
````

And then again to resume the session:

````
openssl s_client -sess_in sess.pem -cert server.pem -key server.pem
````

This resulted in the connection failing and the server reporting:

````
139872190592704:error:140D9115:SSL routines:ssl_get_prev_session:session id context uninitialized:ssl_sess.c:684:
````

So the question is why have you not seen this before, and why are you now seeing it with TLSv1.3. My guess would be related to differences between TLSv1.2 tickets and TLSv1.3 tickets (PSKs). Some theories:

1) You don't tell us anything about your clients for your server. Are they browsers or some other custom written code? Perhaps those clients don't support TLSv1.2 session tickets and therefore they never attempt a ticket based resumption.

2) For some unknown reason your server is not issuing TLSv1.2 tickets, or they are not valid, or they are somehow being marked as non-resumable.

If one of the above guesses is correct, perhaps those conditions are not holding for clients supporting TLSv1.3. In TLSv1.2 session tickets were a later add-on to the protocol so not all clients are required to support them, whereas in TLSv1.3 they are fully integrated. As far as the client is concerned there is no difference between stateful vs stateless tickets - they look the same. So if the client supports resumption at all then it must support tickets (PSKs).

So, it may be the case that (for some as-yet unidentified reason) resumption was either never attempted or never succeeded with your old code/clients, but somehow related to differences in configuration/protocol support resumption now *is* being attempted and *is* succeeding resulting in you seeing this error message for the first time.
